cranelift-wasm/src/code_translator.rs

```diff
@@ -987,18 +987,86 @@ pub fn translate_operator<FE: FuncEnvironment + ?Sized>(
         | Operator::Fence { .. } => {
             return Err(wasm_unsupported!("proposed thread operator {:?}", op));
         }
-        Operator::MemoryInit { .. }
-        | Operator::DataDrop { .. }
-        | Operator::MemoryCopy
-        | Operator::MemoryFill
-        | Operator::TableInit { .. }
-        | Operator::ElemDrop { .. }
-        | Operator::TableCopy
-        | Operator::TableGet { .. }
-        | Operator::TableSet { .. }
-        | Operator::TableGrow { .. }
-        | Operator::TableSize { .. } => {
-            return Err(wasm_unsupported!("proposed bulk memory operator {:?}", op));
+        Operator::MemoryCopy => {
+            // The WebAssembly MVP only supports one linear memory and we assume it here.
+            let heap_index = MemoryIndex::from_u32(0);
+            let heap = state.get_heap(builder.func, 0, environ)?;
+            let len = state.pop1();
+            let src = state.pop1();
+            let dest = state.pop1();
+            environ.translate_memory_copy(builder.cursor(), heap_index, heap, dest, src, len)?;
```

Comment:
Ah, I wasn't aware of those. There are two complications here however.

The first is that the spec requires us to not write anything if the source or destination ranges are OOBs. We can't use the normal bounds checking code because it relies on signal handling to catch accesses that are above `wasm memory.size` but below `cl heap.bound_gv`. Our signal handling won't work inside of the libc::memmove because it ignores faults outside of wasm code. We also wouldn't have a guarantee that no writing would be performed before a fault happened.

If we could access the canonical `wasm memory.size` from the JIT than might be able to do a bounds check against that. The difficulty here is that while the [`size`](https://searchfox.org/mozilla-central/rev/cce8b90aece0f42e5025e45282de16066eeaa662/js/src/wasm/WasmInstance.cpp#368) is at a fixed offset before the heap base, it depends on whether the memory is shared or not. And when the memory is shared, we need to perform an atomic access which doesn't appear to be implemented yet. (although I'm new and may have missed it)

The second complication is that, IIUC, memmove on shared memory is UB in the C/C++ sense. The [implementation](https://searchfox.org/mozilla-central/rev/cce8b90aece0f42e5025e45282de16066eeaa662/js/src/wasm/WasmInstance.cpp#508) in SM uses a JIT'ed version to proactively avoid any issues. Although the original author that introduced this said he might have been too paranoid, so this might not be an issue.

If there is a way around these, that would be great as it would allow other users to share the impl. However this seemed to easiest way to get interoperability with SM.

Reply:
I agree, it makes sense to hand this off to the `environ` here. The `environ` can then decide whether it wants to do `call_memmove` or similar (after checking the bounds).